### ISSUE TITLE
beforeClientResponse is not run

### ISSUE THREAD

**Issue by eriklumme** [mvysny/karibu-testing#11]
It appears to me that `beforeClientResponse` of the current UI is not being run.

While this is expected when running browserless, it would be useful if it could be simulated somehow.

**Comment by mvysny** [comment]
Sure, let's discuss how we can simulate it. It can be called before and after every component lookup perhaps. What kind of functionality do you have hooked there? Can you describe your use case?

**Comment by eriklumme** [comment]
I use it for updating a grid.

The reason is that the grid is in a paging, tabbed view. Depending on the order of operations in terms of changing tab and page, the components may or may not be attached when the update function is called.

Using beforeClientResponse I ensure that my grid is always attached before I try updating it.

**Comment by eriklumme** [comment]
I think that's a good start! In one of my use cases I'd need it to be called after `selectTab` on a tabsheet. I noticed that it in turns calls the `markAsDirty()` method.

```java
@Override
public void markAsDirty() {
    assert getSession() == null
            || getSession().hasLock() : buildLockAssertMessage(
                    "markAsDirty()");
    UI uI = getUI();
    if (uI != null) {
        uI.getConnectorTracker().markDirty(this);
    }
}
```

So that's another option, if Kotlin can do some magic. However, his might not be that straightforward to implement, and I'm not sure if it would be useful for many other use cases than mine.

**Comment by mvysny** [comment]
Hi, I've implementing this by using the `TestingLifecycleHook` and simply calling `UI.getCurrent().internals.stateTree.runExecutionsBeforeClientResponse()` in `awaitBeforeLookup()`. You can actually try to do exactly the same since you can provide your own implementation of `TestingLifecycleHook`. Could you please try and let me know whether it works for your use-case?

**Comment by eriklumme** [comment]
It works if I change my code to `_find` the components after selecting the tab.

I previously had something like
```
grid = _find(Grid.class).get(0);

selectTab(1);

Assert.assertEquals(9000, GridKt._size(grid));
```

While this obviously didn't work, it works when switching order of the first two row, which also is a bit more logical. I don't think running it on `_size` as well would be a good idea.

**Comment by mvysny** [comment]
Thanks! Your code is however a valid use-case and exposes the fact that this fix is too magic, since this:
```kotlin
grid = _find(Grid.class).get(0);
selectTab(1);
Assert.assertEquals(9000, GridKt._size(grid));
```
doesn't work, but this:
```kotlin
selectTab(1);
Assert.assertEquals(9000, GridKt._size(_find(Grid.class).get(0)));
```
does work :(

However, I don't think we can get rid of this magic. So far I've come up with three possible solutions, neither of those is magic-free:

1. Revert the fix and do what Karibu-Testing did previously: immediately execute the block when `beforeClientResponse()` has been called. Apparently that didn't work for you.
2. Go with this solution, accept the magic and clearly explain it in the docs.
3. Create a function `clientRoundtrip()` and expect programmers to call it. I don't think they will :)

Note: you can use `_get()` instead of `_find()` - `_get()` also asserts that there is exactly one Grid which may make your test more solid.

**Comment by eriklumme** [comment]
Magic is bad, yes.

I also thought about 3, but as you say it's likely that users will not remember to call it.

Maybe reverting the fix is the right thing to do. Hopefully the cases where this will cause a major headache for a developer are few and far between.

`_get()` sounds good, I think I should browse through the documentation once more to see what else I missed :)

**Comment by mvysny** [comment]
Thanks for your thoughts!

I think that former code was worse in a sense that instead of obeying the semantics and delaying the block in `beforeClientResponse()` it executed it right away. That could break some code, and I think that's a bug in the Karibu mock-up.

Therefore I still propose to go with this solution and let's hook it up with the `TestingLifecycleHook`. We can then focus to figure out the best way to reproduce the correct browser-client lifecycle in Karibu Testing, then we can hook `TestingLifecycleHook` into that.

**Comment by mvysny** [comment]
Fixed in KT 1.0.3
